core/selection.js

```diff
@@ -22,12 +22,13 @@ class Selection {
     this.composing = false;
     this.mouseDown = false;
     this.root = this.scroll.domNode;
+    this.rootDocument = (this.root.getRootNode ? this.root.getRootNode() : document);
```

Comment:
I'm not sure I like this approach as it creates an extra state and scope dependency. A generalized solution would be much preferred which would also cut down on code duplication in any other modules. Can we use [`ownerDocument`](https://developer.mozilla.org/en-US/docs/Web/API/Node/ownerDocument) instead?

Reply:
afaik `ownerDocument` will always be the owning document, not the shadow root (a document fragment)

Reply:
A possible solution for now:
```javascript
this.rootDocument = this.getDomRoot(this.root)
```
```javascript
  getDomRoot(ctx) {
    if (ctx === document) {
      return ctx
    }
    if (HTMLElement.prototype.attachShadow) {
      if (ctx instanceof ShadowRoot) {

        if (typeof ctx.getSelection === 'function') {
          return ctx
        }

        return document
      }
    }

    return this.getDomRoot(ctx.parentNode)
  }
```
A modified variation from [here](https://github.com/arsnebula/quill/commit/a6489a626f0e5d186ea9114d64efd0a3e1c44186#diff-ae237139107d046c67dc8622d25bace9).